dianna/utils/maskers.py

```diff
@@ -97,57 +96,179 @@ def _get_mask_value(data: np.array, mask_type: str) -> int:
-    if user_requested_steps == series_length:
+    """Determine the number of time steps that need to be masked."""
+    mean = series_length * (1 - p_keep)
+    floor = np.floor(mean)
+    ceil = np.ceil(mean)
+    if floor != ceil:
+        user_requested_steps = int(
+            np.random.choice([floor, ceil], 1, p=[ceil - mean, mean - floor]))
+    else:
+        user_requested_steps = int(floor)
+
+    if user_requested_steps >= series_length:
         warnings.warn(
             'Warning: p_keep chosen too low. Continuing with leaving 1 time step unmasked per mask.'
         )
         return series_length - 1
-    if user_requested_steps == 0:
+    if user_requested_steps <= 0:
         warnings.warn(
             'Warning: p_keep chosen too high. Continuing with masking 1 time step per mask.'
         )
         return 1
     return user_requested_steps
 
 
-def _upscale(grid_i, up_size):
-    return resize(grid_i,
-                  up_size,
-                  order=1,
-                  mode='reflect',
-                  anti_aliasing=False)
+def generate_time_step_masks(input_data: np.ndarray, number_of_masks: int,
+                             p_keep: float, number_of_features: int):
+    """Generate masks that masks complete time steps at a time while masking time steps in a segmented fashion."""
+    time_series_length = input_data.shape[0]
+    number_of_channels = input_data.shape[1]
+
+    float_masks = _generate_interpolated_float_masks_for_timeseries(
+        [time_series_length, 1], number_of_masks, number_of_features)[:, :, 0]
+    bool_masks = np.empty(shape=float_masks.shape, dtype=bool)
+
+    # Convert float masks to bool masks using a dynamic threshold
+    for i in range(float_masks.shape[0]):
+        bool_masks[i] = _mask_bottom_ratio(float_masks[i], p_keep)
 
+    return np.repeat(bool_masks, number_of_channels, axis=2)
 
-def generate_masks_for_images(input_size, n_masks, p_keep, feature_res):
+
+def _mask_bottom_ratio(float_mask: np.ndarray, p_keep: float) -> np.ndarray:
+    """Return a bool mask given a mask of floats and a ratio.
+
+    Return a mask containing bool values where the top p_keep values of the float mask remain unmasked and the rest is
+    masked.
+
+    Args:
+        float_mask: a mask containing float values
+        p_keep: the ratio of keeping cells unmasked
+
+    Returns:
+        a mask containing bool
+    """
+    flat = float_mask.flatten()
+    time_indices = list(range(len(flat)))
+    number_of_unmasked_cells = _determine_number_masked(
+        p_keep, len(time_indices))
+    top_indices = heapq.nsmallest(number_of_unmasked_cells,
+                                  time_indices,
+                                  key=lambda time_step: flat[time_step])
+    flat_mask = np.ones(flat.shape, dtype=bool)
+    flat_mask[top_indices] = False
+    return flat_mask.reshape(float_mask.shape)
+
+
+def _generate_interpolated_float_masks_for_image(input_size: int,
```

Comment:
I couldn't understand the reason why this function is a private function with an underscore in its name. I see that it is used in `rise_image` module. Can we make the name without underscore?

Reply:
Good suggestion. I changed the function to public (removed the underscore).